src/widgets/forms/basic.ts

```diff
@@ -0,0 +1,235 @@
+import { st, Literal, Node, NamedNode } from 'rdflib'
+import store from '../../store'
+import ns from '../../ns'
+import { textInputStyle } from '../../style'
+import { label } from '../../utils'
+
+import { errorMessageBlock } from '../error'
+import { mostSpecificClassURI } from './fieldFunction'
+import { fieldParams } from './fieldParams'
+
+/**
+ * Create an anchor element with a label as the anchor text.
+ *
+ * @param dom The DOM
+ * @param property href for the anchor element
+ * @param fieldInQuestion field to produce a label for
+ *
+ * @ignore exporting this only for unit tests
+ */
+export function fieldLabel (dom: HTMLDocument, property: NamedNode | undefined, fieldInQuestion: Node): HTMLElement | Text {
+  let lab = store.any(fieldInQuestion, ns.ui('label'))
+  if (!lab) lab = label(property, true) // Init capital
+  if (property === undefined) {
+    return dom.createTextNode('@@Internal error: undefined property')
+  }
+  const anchor = dom.createElement('a')
+  /* istanbul ignore next */
+  if (property.uri) anchor.setAttribute('href', property.uri)
+  anchor.setAttribute('style', 'color: #3B5998; text-decoration: none;') // Not too blue and no underline
+  anchor.textContent = lab
+  return anchor
+}
+
+/**
+ * Returns the document for the first quad that matches
+ * the subject and predicate provided, or default if that
+ * store is not editable.
+ *
+ * @param subject Subject about which we want to find an editable RDF document
+ * @param predicate Predicate about which we want to find an editable RDF document
+ * @param def default RDF document to return if none found
+ *
+ * @ignore exporting this only for unit tests
+ */
+export function fieldStore (subject: Node, predicate: Node, def: Node | undefined): Node | undefined {
+  const sts = store.statementsMatching(subject, predicate)
+  if (sts.length === 0) return def // can used default as no data yet
+  if (
+    sts.length > 0 &&
```

Comment:
That's kinda implied by line 47, ain't it?

Reply:
yeah, I didn't do any refactor in this branch. Now that we have full branch coverage, it's definitely something I want to do in a next PR!